Support online and offline expansion without requiring them to be declared (#429)
container-storage-interface/spec: spec.md

    Type type = 1;
  }

  // Deprecated - Ability for a plugin to delcare online and offline
  // controller expansion capabilities via PluginCapability
  // is deprecated. A plugin may support either mode of operation
  // without having to declare them in PluginCapability.
  message VolumeExpansion {
    enum Type {

>       UNKNOWN = 0;

Reviewer: Should we also remove the comments on `ONLINE = 1;` and `OFFLINE = 2;` below and replace them with `// Deprecated` + ` [deprecated=true]` annotation?

  oneof type {
    // Service that the plugin supports.
    Service service = 1;

>     // deprecated

Reviewer 2: "deprecated" should also be added in front of the "message VolumeExpansion" on line 598 as well?

Author: fixed

This operation MUST be idempotent.
If a volume corresponding to the specified volume ID is already larger than or equal to the target capacity of the expansion request, the plugin SHOULD reply 0 OK.

This call MAY be made by the CO during any time in the lifecycle of the volume after creation but an SP may not permit controller expansion of volumes which are controller published or available on a node. In which case - the plugin may return gRPC error code `FAILED_PRECONDITION` (Volume in use) and CO SHOULD ensure that volume is not published and retry with exponential backoff.

> 

Reviewer 2: s/and retry/before retrying

Author: fixed

This operation MUST be idempotent.
If a volume corresponding to the specified volume ID is already larger than or equal to the target capacity of the expansion request, the plugin SHOULD reply 0 OK.

> This call MAY be made by the CO during any time in the lifecycle of the volume after creation but an SP may not permit controller expansion of volumes which are controller published or available on a node. In which case - the plugin may return gRPC error code `FAILED_PRECONDITION` (Volume in use) and CO SHOULD ensure that volume is not published before retrying with exponential backoff.

Reviewer: ```suggestion
This call MAY be made by the CO during any time in the lifecycle of the volume after creation. If an SP does not permit controller expansion of volumes which are controller published or available on a node, the SP may return a gRPC error code `FAILED_PRECONDITION` (Volume in use) and the CO SHOULD ensure that volume is not published on controller or node before retrying with exponential backoff.
```

> 

Reviewer 2: Line 659 says "published and available volumes on a node".  Here it says "published or available on a node".  Is it "and" or "or"?  Should be consistent in both places.

Reviewer: or makes more sense I think

If plugin has `EXPAND_VOLUME` node capability, then `NodeExpandVolume` MUST be called after successful `ControllerExpandVolume` and `node_expansion_required` in `ControllerExpandVolumeResponse` is `true`.

If specified, the `volume_capability` in `ControllerExpandVolumeRequest` should be same as what CO would pass in `ControllerPublishVolumeRequest`.

Examples:

> * Offline Volume Expansion(SP does not permit controller expansion of controller published volume):

Reviewer 2: nit: Add a space before (

> * Offline Volume Expansion(SP does not permit controller expansion of controller published volume):

Reviewer: Should we remove the offline example altogether?

Reviewer 2: I think this example is very important for CSI drivers that support offline volume expansion.  It clearly explains how this case would work.

|-----------|-----------|-------------|-------------------|
| Exceeds capabilities | 3 INVALID_ARGUMENT | Indicates that CO has specified capabilities not supported by the volume. | Caller MAY verify volume capabilities by calling ValidateVolumeCapabilities and retry with matching capabilities. |
| Volume does not exist | 5 NOT FOUND | Indicates that a volume corresponding to the specified volume_id does not exist. | Caller MUST verify that the volume_id is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |

> | Volume in use | 9 FAILED_PRECONDITION | Indicates that the volume corresponding to the specified `volume_id` could not be expanded because it is currently published on a node. | Caller SHOULD ensure that volume is not published and retry with exponential back off. |

Reviewer: ```suggestion
| Volume in use | 9 FAILED_PRECONDITION | Indicates that the volume corresponding to the specified `volume_id` could not be expanded because it is currently published on a controller or node and the SP does not support expansion while published. | Caller SHOULD ensure that volume is not published to controller or node before retrying with exponential back off. |
```
